Minor fixes and changes to XML logging functions

nrwahl2 opened the pull request:
This is in preparation for formatted output in XML logging, which in turn is the last missing piece for `crm_shadow` formatted output.

There are plans to use `pcmk__output_get_log_level()` and `pcmk__text_output_new()`.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -1762,16 +1757,13 @@ log_data_element(int log_level, const char *file, const char *function,
     if (data == NULL) {
-        do_crm_log_alias(log_level, file, function, line, "%s: %s", prefix,
-                         "No data to dump as XML");
+        do_crm_log(log_level, "%s: %s", prefix, "No data to dump as XML");
```
This format string doesn't make much sense to me if `prefix` is the empty string or contains only spaces.

kgaillot replied:
Yeah, I think we just never end up in this function with data==NULL. To make it even more fun, some callers of `crm_log_xml_*` use a ": " in the prefix, and some bracket the prefix.

The strange usage was added in 37161114 -- previously it was just a crm_warn() without the prefix. The comment about libqb no longer applies since we're not logging under an alias anymore.

I'm thinking we should drop the ": " here. We could be more consistent with the `crm_log_xml_*()` callers as far as brackets or whatnot, but that's not important now.

nrwahl2 commented on lib/common/xml.c:
```diff
@@ -1754,7 +1754,7 @@ log_data_element(int log_level, const char *file, const char *function,
 
     if (prefix == NULL) {
-        prefix = "";
+        prefix = "  ";
```
This is really a guess, so I'd appreciate some insight here. Are the various prefixes and their meanings documented anywhere? There are a bunch of `+`, `-`, and `~` symbol combinations scattered throughout the patchset/xml code.

Below here, we set the element at index 1 to either `'+'` or `'-'` if particular flags are set, after defaulting `prefix` to an empty string -- so that's an out-of-bounds access. This is an attempt to fix that issue. But... we're only modifying the second element below. What's supposed to be in the first element?

For example, if the prefix is some arbitrary string, it doesn't make sense to overwrite the second character. And if the first character is a plus, I'm not sure if it makes sense for the second character to be a minus.

kgaillot replied:
> This is really a guess, so I'd appreciate some insight here. Are the various prefixes and their meanings documented anywhere? There are a bunch of `+`, `-`, and `~` symbol combinations scattered throughout the patchset/xml code.

Undocumented, but I believe `--` indicates something removed from the original CIB, `++` indicates something added or replaced, and `+~` indicates something moved. `- ` or `+ ` identifies the original or patched CIB somehow.
 
> Below here, we set the element at index 1 to either `'+'` or `'-'` if particular flags are set, after defaulting `prefix` to an empty string -- so that's an out-of-bounds access. This is an attempt to fix that issue. But... we're only modifying the second element below. What's supposed to be in the first element?

I suspect prefix never being empty is the only thing avoiding a segfault. We could CRM_CHECK() it.

> For example, if the prefix is some arbitrary string, it doesn't make sense to overwrite the second character. And if the first character is a plus, I'm not sure if it makes sense for the second character to be a minus.

I suspect `--`, `++`, and `+~` are the only possible two-character combinations in practice.

nrwahl2 replied:
> Undocumented, but I believe -- indicates something removed from the original CIB, `++` indicates something added or replaced, and `+~` indicates something moved. `-` or `+` identifies the original or patched CIB somehow.

That sounds right... it looks like callers often pass in `"+ "` or `"- "` with the respective `xml_log_option_diff_*` option. Then if `xml_log_option_dirty_add` is not set, `log_data_element` overwrites the space with the second `'+'` or `'-'` character. If `xml_log_option_dirty_add` is set, then it calls `log_xml_changes()`, which sets the first two characters to `"++"`, `"--"`, or `"+~"`.

It might be worth redoing it a little bit more clearly but I'm not sure when I'm gonna want to do that.

> I suspect prefix never being empty is the only thing avoiding a segfault. We could CRM_CHECK() it.

That'll work.

> I suspect `--`, `++`, and `+~` are the only possible two-character combinations in practice.

The only meaningful ones at least. For example, `do_crm_log_xml()` passes an arbitrary string as the prefix. That's part of what I was worried about -- overwriting the first and/or second character. I think we're safe in that case because the options flag group is hard-coded as `xml_log_option_formatted`, so we won't have any of the diff or dirty_add flags set.

kgaillot commented on lib/common/output_log.c:
```diff
@@ -329,6 +329,18 @@ pcmk__mk_log_output(char **argv) {
 }
 
+int
+pcmk__output_get_log_level(pcmk__output_t *out)
```
could be const

nrwahl2 replied:
Yep thanks. I'm gonna pull this commit out of this PR, in case I don't end up using it. I'm 90% sure I will, but the formatted output of XML and patchsets is messy.

kgaillot reviewed (commented):
looks good

nrwahl2 commented on lib/common/output_xml.c:
```diff
@@ -453,16 +453,21 @@ pcmk__output_xml_create_parent(pcmk__output_t *out, const char *name, ...) {
+    parent = g_queue_peek_tail(priv->parent_q);
+
+    // Shouldn't happen unless the caller popped priv->root
+    CRM_CHECK(parent == NULL, return);
```
Whoops, yeah that's a big one. No caller as of this PR so it went undetected by me.

nrwahl2 commented:
Fixed `CRM_CHECK` bug per review, removed `pcmk__output_get_log_level()` from this PR on the off chance it doesn't get used, and rebased on current main.

nrwahl2 commented:
Scrolled down too far and missed the thread about `CRM_CHECK`-ing `prefix`. Updated with that.

About to add a commit on top to remove the `": "` in the `do_crm_log()` line. That will look a little bit weird if the prefix is of an `"[+-] "` form, but that should be rare or impossible and would look weird even with `": "`.

A lot of this weirdness comes from a sort of overloading of `prefix`. In some cases, it's an arbitrary text string, like a descriptive identifier for the XML element. In other cases, it's a semantically meaningful part of an XML diff (e.g., `"+ "`). I don't think it's worth the effort to separate those uses now, especially with some of the affected functions being public, but it is annoying.
